### ISSUE TITLE
Global (whitespace) toggle in hooks

### ISSUE THREAD

**Issue by truchi** [syl20bnr/spacemacs#12136]
*Hello good sirs*,

**tl;dr**
**Is spacemacs/toggle-whitespace-globally-[on|off] working when hooking?**

There is something I do not understand about the global whitespace toggle, it may be a bug.

It is initially on. I am trying to disable it using hooks when:
- in visual mode,
- company is showing a popup.

My hooks are not doing anything.

I will present you 5 (very similar) use-cases to demonstrate the problem:

| Use-case | Initially | Hooks | Works? |
| - | - | - | - |
| 1 - WS | global | global | NO |
| 2 - HLL | global | global | YES |
| 3 - WS | global | non global | NO |
| 4 - HLL | global | non global | YES&ast; |
| 5 - WS | non global | non global | YES |

(WS: whitespace, HLL: highlight long lines (for demonstration/investigation purposes), &ast;: see below)
(All code snippets in `spacemacs.el` under `user/config`, I `SPC q r` before testing)

***

- OS: Ubuntu 16.04
- Emacs version: GNU 24.5.1
- Spacemacs version: v.0.200.13
- Enabled layers: html, javascript, react, helm, auto-completion, better-defaults, emacs-lisp, git, syntax-checking, version-control.
- Excluded packages: fill-column-indicator, evil-escape.

***

### 1 - Whitespace globally doesn't work...

The following doesn't work (what I am trying to achieve, see also 3).
Whitespace is globally on at startup indeed, but going in visual or company doesn't toggle it off (but changes color and weight for some reason).

```elisp
  (spacemacs/toggle-whitespace-globally-on)

  ;; NO WHITESPACE IN VISUAL & COMPANY -- START
  (defun my-toggle-whitespace-off (&optional arg)
    (spacemacs/toggle-whitespace-globally-off))
  (defun my-toggle-whitespace-on (&optional arg)
    (spacemacs/toggle-whitespace-globally-on))

  (add-hook 'company-completion-started-hook 'my-toggle-whitespace-off)
  (add-hook 'company-after-completion-hook   'my-toggle-whitespace-on)
  (add-hook 'evil-visual-state-entry-hook    'my-toggle-whitespace-off)
  (add-hook 'evil-visual-state-exit-hook     'my-toggle-whitespace-on)
  ;; NO WHITESPACE IN VISUAL & COMPANY -- END
```

### 2 - ...but highlight long lines does!

The same thing with the highlight long lines toggle works fine.

```elisp
  (spacemacs/toggle-highlight-long-lines-globally-on)

  ;; NO HIGHLIGHT LONG LINES IN VISUAL & COMPANY -- START
  (defun my-toggle-highlight-long-lines-off (&optional arg)
    (spacemacs/toggle-highlight-long-lines-globally-off))
  (defun my-toggle-highlight-long-lines-on (&optional arg)
    (spacemacs/toggle-highlight-long-lines-globally-on))

  (add-hook 'company-completion-started-hook 'my-toggle-highlight-long-lines-off)
  (add-hook 'company-after-completion-hook   'my-toggle-highlight-long-lines-on)
  (add-hook 'evil-visual-state-entry-hook    'my-toggle-highlight-long-lines-off)
  (add-hook 'evil-visual-state-exit-hook     'my-toggle-highlight-long-lines-on)
  ;; NO HIGHLIGHT LONG LINES IN VISUAL & COMPANY -- END

```

### 3 - Still not working when toggling non globally

(Also what I am trying to achieve in a different - probably more idiomatic - way)
Not working either when toggling *non globally* in the hook callbacks with *globally* on initially.

```elisp
  (spacemacs/toggle-whitespace-globally-on)

  ;; NO WHITESPACE IN VISUAL & COMPANY -- START
  (defun my-toggle-whitespace-off (&optional arg)
    (spacemacs/toggle-whitespace-off))
  (defun my-toggle-whitespace-on (&optional arg)
    (spacemacs/toggle-whitespace-on))

  (add-hook 'company-completion-started-hook 'my-toggle-whitespace-off)
  (add-hook 'company-after-completion-hook   'my-toggle-whitespace-on)
  (add-hook 'evil-visual-state-entry-hook    'my-toggle-whitespace-off)
  (add-hook 'evil-visual-state-exit-hook     'my-toggle-whitespace-on)
  ;; NO WHITESPACE IN VISUAL & COMPANY -- END
```

### 4 - Same with highlight long lines works (kinda)

The same pattern with highlight long lines works in visual.
Completions popup breaks (company: auto-begin: wrong number of arguments: ...). This is probably due to my noobish skills in elisp.

```elisp
  (spacemacs/toggle-highlight-long-lines-globally-on)

  ;; NO HIGHLIGHT LONG LINES IN VISUAL & COMPANY -- START
  (defun my-toggle-highlight-long-lines-off (&optional arg)
    (spacemacs/toggle-highlight-long-lines-off))
  (defun my-toggle-highlight-long-lines-on (&optional arg)
    (spacemacs/toggle-highlight-long-lines-on))

  (add-hook 'company-completion-started-hook 'my-toggle-highlight-long-lines-off)
  (add-hook 'company-after-completion-hook   'my-toggle-highlight-long-lines-on)
  (add-hook 'evil-visual-state-entry-hook    'my-toggle-highlight-long-lines-off)
  (add-hook 'evil-visual-state-exit-hook     'my-toggle-highlight-long-lines-on)
  ;; NO HIGHLIGHT LONG LINES IN VISUAL & COMPANY -- END
```

### 5 - Falling back to non global

As a fallback I will use the following. Works, except I have no whitespaces initially.

```elisp
  (spacemacs/toggle-whitespace-on)

  ;; NO WHITESPACE IN VISUAL & COMPANY -- START
  (defun my-toggle-whitespace-off (&optional arg)
    (spacemacs/toggle-whitespace-off))
  (defun my-toggle-whitespace-on (&optional arg)
    (spacemacs/toggle-whitespace-on))

  (add-hook 'company-completion-started-hook 'my-toggle-whitespace-off)
  (add-hook 'company-after-completion-hook   'my-toggle-whitespace-on)
  (add-hook 'evil-visual-state-entry-hook    'my-toggle-whitespace-off)
  (add-hook 'evil-visual-state-exit-hook     'my-toggle-whitespace-on)
  ;; NO WHITESPACE IN VISUAL & COMPANY -- END
```

***

Is this a bug ?
Am I doing something wrong ?

Thank you very much in advance for your time and efforts.
